### PR TITLE
Sample support for index generator

### PR DESCRIPTION
**Pull request by GeekArthur**
Signed-off-by: jingfu wang <jingfwan@redhat.com>

**Please specify the area for this PR**
index generator

**What does does this PR do / why we need it**:
This PR modifies the index generator to let it parse extraDevfileEntires.yaml and generate index.json with both stacks and samples

**Which issue(s) this PR fixes**:

Related issues:
https://github.com/devfile/api/issues/279
https://github.com/devfile/api/issues/403
https://github.com/devfile/api/issues/409

**PR acceptance criteria**:

- [x] Test (WIP) 

- [x] Documentation (WIP)

**How to test changes / Special notes to the reviewer**:
1. Build the index generator
2. Run the binary to generate index.json then verify it (Note: the registry path is the registry root directory)

### CODE REVIEW

**Review by johnmcollier — CHANGES_REQUESTED** [review]
Left a couple comments.

File: index/generator/library/library.go [hunk 1]
```diff
@@ -89,6 +94,39 @@ func GenerateIndexStruct(registryDirPath string, force bool) ([]schema.Schema, e
 		index = append(index, indexComponent)
 	}
 
+	// Parse extraDevfileEntries.yaml
```
[the remark this answers is not in the document]
Comment (GeekArthur) [reply index/generator/library/library.go:97]: Sure, I can make the changes and add unit tests.

[the remark this answers is not in the document]
Comment (GeekArthur) [reply index/generator/library/library.go]: Done

File: index/generator/library/library.go [hunk 2]
```diff
@@ -107,15 +145,21 @@ func CreateIndexFile(index []schema.Schema, indexFilePath string) error {
...
-func validateIndexComponent(indexComponent schema.Schema) error {
-	if indexComponent.Name == "" {
-		return fmt.Errorf("index component name is not initialized")
-	}
-	if indexComponent.Links == nil {
-		return fmt.Errorf("index component links are empty")
-	}
-	if indexComponent.Resources == nil {
-		return fmt.Errorf("index component resources are empty")
+func validateIndexComponent(indexComponent schema.Schema, componentType string) error {
```
[the remark this answers is not in the document]
Comment (GeekArthur) [reply index/generator/library/library.go]: Done

**Comment by GeekArthur** [comment]
@johnmcollier Thanks for reviewing, I refactored the functions and added the unit tests, please review again.

**Review by johnmcollier — APPROVED** [review]
Looks good, great job on the tests too 👍

**Pull request merged**